File: kernel/drivers/agnostic/serialfs.lua
local BaseFS = require("drivers.common.basefs")
local TwoWayMap = require("common.two_way_map")

---@class SerialFS: BaseFS
---@field host_path string
---@field data table<integer, string>
local SerialFS = BaseFS:new()

---@param arch Arch
---@param host_path string  path on the host filesystem to persist to
---@return SerialFS
function SerialFS:new(arch, host_path)
    local fs = BaseFS.new(self, arch) --[[@as SerialFS]]
    fs.host_path = host_path
    fs.data = {}
    return fs
end

-- Minimal serializer: handles strings, numbers, booleans, and nested tables.
local function serialize(val, indent)
    indent = indent or ""
    local t = type(val)
    if t == "string" then
        return string.format("%q", val)
    elseif t == "number" or t == "boolean" then
        return tostring(val)
    elseif t == "table" then
        local parts = {}
        local ni = indent .. "  "
        for k, v in pairs(val) do
            local key = type(k) == "string" and k:match("^[%a_][%w_]*$")
                and k or ("[" .. serialize(k) .. "]")
            parts[#parts + 1] = ni .. key .. " = " .. serialize(v, ni)
        end
        if #parts == 0 then return "{}" end
        return "{\n" .. table.concat(parts, ",\n") .. "\n" .. indent .. "}"
    else
        return "nil"
    end
end

function SerialFS:_save()
    local f = assert(io.open(self.host_path, "w"))
    f:write(serialize({
        inode_id = self.inode_id,
        inodes   = self.inodes,
        paths    = self.inode_path_map.forward,
        data     = self.data,
    }))
    f:close()
end

local function _load(host_path)
    local f = io.open(host_path, "r")
    if not f then return nil end
    local content = f:read("*a")
    f:close()
    local fn = load("return " .. content)
    return fn and fn()
end

-- Override mount: restore state from file if it exists, otherwise initialize fresh.
---@param resolved_path string
---@return Inode
function SerialFS:mount(resolved_path)
    local state = _load(self.host_path)
    if state then
        self.inode_id = state.inode_id
        self.inodes   = state.inodes
        self.data     = state.data
        local map = TwoWayMap:new()
        for path, id in pairs(state.paths) do
            map:set(path, id)
        end
        self.inode_path_map = map
        return self.inodes[self.inode_path_map:get(resolved_path)]
    else
        local inode = BaseFS.mount(self, resolved_path)
        self:_save()
        return inode
    end
end

---@param parent_inode Inode
---@param name string
---@param type integer
---@return Inode
function SerialFS:create_file(parent_inode, name, type)
    local inode = BaseFS.create_file(self, parent_inode, name, type)
    self:_save()
    return inode
end

---@param inode Inode
function SerialFS:destroy_file(inode)
    BaseFS.destroy_file(self, inode)
    if not inode.unlinked then
        -- Inode was freed immediately (no open fds); clean up data now.
        self.data[inode.id] = nil
    end
    self:_save()
end

-- Called by FileDescriptor:close() when the last fd on an unlinked inode closes.
---@param inode Inode
function SerialFS:_free_inode(inode)
    BaseFS._free_inode(self, inode)
    self.data[inode.id] = nil
    self:_save()
end

---@param inode Inode
---@param offset integer
---@param length integer
function SerialFS:read_file(inode, offset, length)
    local contents = self.data[inode.id] or ""
    return contents:sub(offset + 1, offset + length)
end

---@param inode Inode
---@param offset integer
---@param data any
function SerialFS:write_file(inode, offset, data)
    local contents = self.data[inode.id] or ""
    local prefix = contents:sub(1, offset)
    if #prefix < offset then
        prefix = prefix .. string.rep("\0", offset - #prefix)
    end
    local suffix = contents:sub(offset + #data + 1)
    self.data[inode.id] = prefix .. data .. suffix
    inode.size = #self.data[inode.id]
    self:_save()
    return #data
end

return SerialFS

File: kernel/drivers/agnostic/serialfs.ts
import { BiMap } from "../../common/bimap";
import { Inode } from "../../vfs/inode/init";
import { BaseFS } from "../common/basefs";
import type { Arch } from "../../arch/arch";

function serialize(value: unknown, indent: string = ""): string {
	let t = type(value);

	switch(t) {
		case "string":
			return string.format("%q", value);
		case "number":
		case "boolean":
			return tostring(value);
		case "table":
			let parts: string[] = [];
			let ni = indent + "  ";
			for (const [k, v] of pairs(value)) {
				// Bare identifier keys can be written unquoted; everything else
				// (paths with "/", numeric keys) must be bracketed+quoted. Note
				// the match is destructured to a single value: used directly in
				// a boolean, tstl wraps the multi-return in a table that is
				// always truthy, which would emit paths unquoted.
				let key: string;
				if (typeof(k) === "string") {
					let [ident] = string.match(k, "^[%a_][%w_]*$");
					key = ident !== undefined ? k : "[" + serialize(k) + "]";
				} else {
					key = "[" + serialize(k) + "]";
				}
				parts.push(ni + key + "=" + serialize(v, ni));
			}
			if (parts.length === 0) { return "{}" };
			return "{\n" + parts.join(",\n") + "\n" + indent + "}";
		default:
			return "nil"
	}
}

export class SerialFS extends BaseFS {
	host_path: Path;
	data: Map<integer, string> = new Map();

	constructor(arch: Arch, host_path: Path) {
		super(arch);
		this.host_path = host_path;
	}

	private save() {
		let [f] = io.open(this.host_path, "w");
		if (f === undefined) { return; }

		// On-disk layout is a plain, tool-friendly format (see
		// build/mkserialfs.lua): inodes and data keyed by inode id, paths as
		// path -> id. Convert the in-memory tstl structures (a 1-based array,
		// a Map and a BiMap) into that plain form so the file round-trips and
		// stays compatible with the build-time packer.
		let inodes = new LuaTable<integer, Inode>();
		for (const [_, inode] of pairs(this.inodes as unknown as LuaTable<integer, Inode>)) {
			inodes.set(inode.id, inode);
		}
		let data = new LuaTable<integer, string>();
		for (const [id, content] of this.data) {
			data.set(id, content);
		}
		let paths = new LuaTable<Path, integer>();
		for (const [path, id] of this.inode_path_map.entries()) {
			paths.set(path, id);
		}

		f.write(serialize({
			inode_id: this.inode_id,
			inodes: inodes,
			paths: paths,
			data: data
		}));
		f.close();
	}

	private static load(host_path: Path): LuaTable | undefined {
		let [f] = io.open(host_path, "r");
		if (f === undefined) { return undefined; }
		let content = f.read("*a");
		f.close();
		let fn = load("return " + content)[0];
		if (fn === undefined) { return undefined; }
		return fn();
	}

	mount(resolved_path: Path): Inode {
		let state = SerialFS.load(this.host_path);
		if (state !== undefined) {
			this.inode_id = state.get("inode_id");

			// Rebuild the in-memory structures from the plain on-disk layout.
			// Assigning through the typed fields lets tstl apply its array
			// (1-based) and Map representations; a raw assignment of the disk
			// tables would leave inodes off-by-one and data as a non-Map.
			let inodes: Inode[] = [];
			for (const [id, inode] of pairs(state.get("inodes") as LuaTable<integer, Inode>)) {
				inodes[id] = inode;
			}
			this.inodes = inodes;

			let data = new Map<integer, string>();
			for (const [id, content] of pairs(state.get("data") as LuaTable<integer, string>)) {
				data.set(id, content);
			}
			this.data = data;

			let map = new BiMap<Path, integer>();
			for (const [path, id] of pairs(state.get("paths") as LuaTable<Path, integer>)) {
				if (typeof(path) !== "string") { throw "Attempted to load malformed serial file"; }
				map.set(path, id);
			}
			this.inode_path_map = map;
			let returned_inode = this.inode_path_map.get(resolved_path);
			if (returned_inode === undefined) { throw "idek this is getting depricated"; }
			return this.inodes[returned_inode];
		} else {
			let inode = super.mount(resolved_path);
			this.save();
			return inode;
		}
	}

	create_file(parent_inode: Inode, name: string, type: InodeModeFlags): Inode {
		let inode = super.create_file(parent_inode, name, type);
		this.save();
		return inode;
	}

	destroy_file(inode: Inode): void {
		super.destroy_file(inode);
		if (!inode.unlinked) {
			this.data.delete(inode.id)
		}
		this.save();
	}

	free_inode(inode: Inode): void {
		super.free_inode(inode);
		this.data.delete(inode.id);
		this.save();
	}

	read_file(inode: globalThis.Inode, offset: integer, length: integer): string {
		let contents = this.data.get(inode.id) || "";
		return string.sub(contents, offset + 1, offset + length);
	}

	write_file(inode: globalThis.Inode, offset: integer, data: string): integer {
		let contents = this.data.get(inode.id) || "";
		let prefix = string.sub(contents, offset);
		if (prefix.length < offset) {
			prefix = prefix + string.rep("\0", offset - prefix.length);
		}
		let suffix = string.sub(contents, offset + data.length + 1);
		let new_contents = prefix + data + suffix;
		this.data.set(inode.id, new_contents);
		inode.size = new_contents.length;
		this.save();
		return data.length;
	}

	ioctl(inode: globalThis.Inode, request: integer, arg: any): unknown {
		return;
	}
}
